Protocols.playground/Contents.swift

```diff
@@ -0,0 +1,648 @@
+        if (currentCargoHeight - cargoHeight < 0) {
+            print("I can't unload more than \(currentCargoHeight) tons of cargo!")
+            return false
+        }
+
+        currentCargoHeight -= cargoHeight
+        return true
+    }
+    
+    mutating func openCargoCompartment() {
+        isCargoСompartmentOpen = true
+    }
+
+    mutating func closeCargoСompartment() {
+        isCargoСompartmentOpen = false
+    }
+}
+
+
+struct Truck: Car, FreightTransport {
+
+    var engine: Engine
+    var driveUnit: DriveUnit
+    var mileage: Double
+    var fuel: Double
+    var manufactorer: Factory
+    var maxFuel: Double
+    var steeringWheel: SteeringWheel
+    var wheelbase: Wheelbase
+    var cargoCapacity: Float
+    var currentCargoHeight: Float
+    var isCargoСompartmentOpen: Bool
+
+    init(engine: Engine, driveUnit: DriveUnit, maxFuel: Double, manufactorer: Factory, wheelType: Wheel, wheelAxlesCount: Int, cargoCapacity: Float) {
+        self.engine = engine
+        self.driveUnit = driveUnit
+        self.mileage = 0
+        self.fuel = 0
+        self.maxFuel = maxFuel
+        self.manufactorer = manufactorer
+        self.steeringWheel = SteeringWheel()
+        self.wheelbase = Wheelbase(axlesCount: wheelAxlesCount, wheel: wheelType)
+        self.isCargoСompartmentOpen = false
+        self.currentCargoHeight = 0
+        self.cargoCapacity = cargoCapacity
+    }
+
+    mutating func transportCargo(cargoWeight: Double, distance: Double) {
+        if move(distance: distance + cargoWeight / 5) {
+            print("I'm strong! I'm transport cargo weighing \(cargoWeight) tons!")
+        }
+    }
+}
+
+
+enum CarType: String {
+    case sedan
+    case offroad
+    case bus
+    case truck
+}
+
+
+protocol CarStorageDelegate: AnyObject {
+    func store(sedans: [Sedan])
+    func store(offroads: [Offroad])
+    func store(buses: [Bus])
+    func store(trucks: [Truck])
+}
+
+
+class Factory {
+
+    private(set) var carCounter: Int
+    private(set) var name: String
+
+    weak var storeDelegate: CarStorageDelegate?
+
+    init(name: String) {
+        self.name = name
+        self.carCounter = 0
+    }
+    
+    private func store(sedans: [Sedan]) {
+        storeDelegate?.store(sedans: sedans)
+    }
+
+    private func store(offroads: [Offroad]) {
+        storeDelegate?.store(offroads: offroads)
+    }
+    
+    private func store(buses: [Bus]) {
+        storeDelegate?.store(buses: buses)
+    }
+    
+    private func store(trucks: [Truck]) {
+        storeDelegate?.store(trucks: trucks)
+    }
+
+    private func createSedans(engine: Engine, driveUnit: DriveUnit, maxFuel: Double, count: Int) {
+        var createdSedans: [Sedan] = []
+
+        for _ in 1...count {
+            createdSedans.append(
+                Sedan(engine: engine, driveUnit: driveUnit, maxFuel: maxFuel, manufactorer: self, wheelType: Wheel(radius: 16, width: 6.5))
+            )
+        }
+
+        store(sedans: createdSedans)
```

Comment:
здесь можно сразу вызвать метод у делегата
```suggestion
    storeDelegate?.store(sedans: sedans)
```